// (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.

syntax = "proto3";

package agent_bus;

// Notes on AgentBus Design:
// Each AgentBus is a physical log of BusEntries, which have different types:
// 1. Intention: a command that an agent wants to (but has not yet) run.
// 2. Vote: an opinion by an agent on whether a command is safe to run.
// 3. DeciderPolicy: a policy for how the Decider aggregates votes on a command.
// 4. Commit: a command that has been committed to run.
// 5. Abort: a command that has been aborted.

// MB: TODO:
// 1. Add support for asynchronous upcalls / streaming apply
// 2. Add support for create/destroy agentbus

// Type of vote info generated by an LLM Voter
message ExternalLlmVoteInfo {
  string reason = 1; // reasoning provided by the LLM for the vote
  string model = 2; // model used to generate the vote
}

// Information about a vote
message VoteInfo {
  oneof vote_info {
    ExternalLlmVoteInfo external_llm_vote_info = 1; // LLM vote info
  }
}

// Enum for filtering payload types in poll requests
enum SelectivePollType {
  SELECTIVE_POLL_TYPE_UNSPECIFIED = 0;
  INTENTION = 1;
  VOTE = 2;
  DECIDER_POLICY = 3;
  COMMIT = 4;
  ABORT = 5;
  VOTER_POLICY = 6;
  CONTROL = 7;
  INFERENCE_INPUT = 8;
  INFERENCE_OUTPUT = 9;
  ACTION_OUTPUT = 10;
  AGENT_INPUT = 11;
  AGENT_OUTPUT = 12;
}

// Types of entries in the AgentBus
message Payload {
  oneof payload {
    Intention intention = 1; // an intented action that requires votes
    Vote vote = 2; // a vote on an intention
    DeciderPolicy decider_policy = 3; // policy for how Decider makes decisions
    Commit commit = 4; // a commit of an intention
    Abort abort = 5; // an abort of an intention
    VoterPolicy voter_policy = 6; // policy for voter customization
    Control control = 7; // a control signal for the bus
    InferenceInput inference_input = 8; // LLM inference input
    InferenceOutput inference_output = 9; // LLM inference output
    ActionOutput action_output = 10; // output from an action (see intention)
    AgentInput agent_input = 11; // input to the agent
    AgentOutput agent_output = 12; // output from the agent
  }
}

// A commit of an intention
message Commit {
  int64 intention_id = 1; // identifies the intention
  string reason = 2; // reason for the commit
}

// An abort of an intention
message Abort {
  int64 intention_id = 1; // identifies the intention
  string reason = 2; // reason for the abort
}

// Types of Control signals in the AgentBus
message Control {
  oneof control {
    string agent_input = 1; // input to the agent
  }
}

// Types of Intentions in the AgentBus
message Intention {
  oneof intention {
    string string_intention = 1; // string-based intention
  }
}

// LLM inference input
message InferenceInput {
  oneof inference_input {
    string string_inference_input = 1; // string-based inference input
  }
}

// LLM inference output
message InferenceOutput {
  oneof inference_output {
    string string_inference_output = 1; // string-based inference output
  }
}

// Output from an action
message ActionOutput {
  int64 intention_id =
      1; // identifies the intention, which links it to the corresponding action
  oneof action_output {
    string string_action_output =
        2; // serialized string representation of the action output/results
  }
}

// Input to the agent (starts an agentic loop)
message AgentInput {
  oneof agent_input {
    string string_agent_input =
        1; // string-based agent input (e.g., user request)
  }
}

// Output from the agent (ends an agentic loop)
message AgentOutput {
  oneof agent_output {
    string string_agent_output =
        1; // string-based agent output (e.g., final response)
  }
}

// A vote on an intention
message Vote {
  VoteType abstract_vote = 1; // abstract vote type
  int64 intention_id = 2; // identifies the intention
  VoteInfo info = 3; // information about the vote
}

// Types of votes in the AgentBus
message VoteType {
  oneof vote_type {
    bool boolean_vote = 1; // boolean
    double probability_vote = 2; // double in [0,1]
  }
}

// Decider Policies for the AgentBus
// Controls how the Decider aggregates votes to make commit/abort decisions
enum DeciderPolicy {
  OFF_BY_DEFAULT = 0;
  ON_BY_DEFAULT = 1;
  FIRST_BOOLEAN_WINS = 2;
}

// Voter Policy for the AgentBus
// Allows customization of voter behavior
message VoterPolicy {
  string prompt_override =
      1; // custom prompt to override the voter's default prompt
}

// Physical Entry in the Bus
// Each entry has a header and a payload.
message BusEntry {
  Header header = 1;
  Payload payload = 2;
}

// Header for a BusEntry
message Header {
  int64 log_position = 1; // physical log position in the AgentBus
}

message ProposeRequest {
  string agent_bus_id = 1; // id for the agent bus
  Payload payload = 2; // payload to be stored
}

message ProposeResponse {
  int64 log_position = 1; // log position where the entry was stored
}

message PollRequest {
  string agent_bus_id = 1; // id for the agent bus
  int64 start_log_position = 2; // inclusive log position to start polling from
  int32 max_entries = 3; // max number of entries to return
  // Optional filter by payload types. If not set, returns all types.
  // If set to empty list, returns no entries.
  optional PayloadTypeFilter filter = 4;
}

message PayloadTypeFilter {
  repeated SelectivePollType payload_types = 1;
}

// MB: TODO: each poll request uses a new linearization point;
// for pagination we may want to use a prior established linearization point
// for later requests
message PollResponse {
  repeated BusEntry entries = 1;
  bool complete = 2; // true if there are no more entries to poll
}

/**
 * AgentBus service interface
 */
service AgentBusService {
  // Propose a command to the agent bus
  // Once this returns, we are guaranteed that
  // the command is durable;
  // and that we will eventually see it via polling
  // (or via a notification, once we support that)
  // Note that a propose can be used for intentions, votes,
  // decider/voter policies, commits, or aborts.
  rpc Propose(ProposeRequest) returns (ProposeResponse);

  // Poll for physical entries in the agent bus
  // Optionally filter by payload types using payloadTypes field
  rpc Poll(PollRequest) returns (PollResponse);
}
